### PR TITLE
fix: ensure account deployment for executTxn in get-starknet 

### PR DESCRIPTION
**Pull request by khanti42**

### CODE REVIEW

File: packages/starknet-snap/src/executeTxn.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/executeTxn.ts:69]: 2 possible issues will making an error

1. when createAccount success but pending in mem pool, in that case the acc is not yet deploy, and the next txn should throw an error saying contract not found/deploy
2. when createAccount success and land to L2, however , a racing condition happening, the next txn is using an old nonce (it is a bug from starknet.js, as we already specific the nonce 1), the error should look like "nonce repreated"

File: packages/starknet-snap/snap.manifest.json
Comment (stanleyyconsensys) [inline packages/starknet-snap/snap.manifest.json:30]: please rollback to false, true means your snap need to talk to other snap, but we dont have such use case

File: packages/starknet-snap/src/executeTxn.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/executeTxn.ts:18]: it looks like we can only support 1 txn

File: packages/starknet-snap/src/executeTxn.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/src/executeTxn.ts:30]: in that case we may use estimateFees 

and build the Invocations in this file

1. init a Invocations base on the request
1. check if account deployed
2. if not , then append left in Invocations for create account
3. run estimateFeeBulk , i think we really dont need to call the estimateFees
4. run addFeesFromAllTransactions(estimateBulkFeeResp); , which can sum the fee for you
5. if yes, just run estimateFeeUtil, again i think we really dont need to call the estimateFees
6. with the fee, we add the gas fee in the snapComponents
7. and display the dialog
8. and process the txn

**Review by stanleyyconsensys — CHANGES_REQUESTED** [review]
LGTM , but we all forgot 1 use case -> the execute txn is allow multiple txn

File: packages/starknet-snap/src/executeTxn.ts
Comment (khanti42) [reply packages/starknet-snap/src/executeTxn.ts]: I adapted the code to run estimateFeeBulk directly from executeTxn.

File: packages/starknet-snap/src/executeTxn.ts
Comment (khanti42) [reply packages/starknet-snap/src/executeTxn.ts]: The code should now work correctly with multiple transactions

**Comment by khanti42** [comment]
should be good for second review @stanleyyconsensys

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:64]: can we have a test case where isAccountDeployed is true

**Review by stanleyyconsensys — CHANGES_REQUESTED** [review]
LGTM, i would like to see if we can have a new test case for if acc deployed, for now we only has a test case where acc not deployed

**Review by stanleyyconsensys — APPROVED** [review]

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:66]:
```suggestion
```

we can remove this, as we wanna to test if create account has triggered

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:113]:
```suggestion
    const stub = sandbox.stub(utils, 'executeTxn').resolves({
      transaction_hash: 'transaction_hash',
    });
    const createAccountStub = sandbox.stub(createAccountUtils, 'createAccount').resolvesThis();
    const result = await executeTxn(apiParams);
    const { privateKey } = await utils.getKeysFromAddress(
      apiParams.keyDeriver,
      STARKNET_MAINNET_NETWORK,
      state,
      account1.address,
    );

    expect(result).to.eql({
      transaction_hash: 'transaction_hash',
    });
    expect(stub).to.have.been.calledOnceWith(
      STARKNET_MAINNET_NETWORK,
      account1.address,
      privateKey,
      {
        entrypoint: 'transfer',
        calldata: ['0', '0', '0'],
        contractAddress: createAccountProxyTxn.contractAddress,
      },
      undefined,
      { maxFee: '22702500105945', nonce: 1 },
    );
    expect(createAccountStub).to.have.been.calledOnceWith(sinon.match.any, true, true);
```

we check if createAccountStub has triggered or not

remember to include import * as createAccountUtils from '../../src/createAccount'; at the top if the file

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:117]:
```suggestion
    sandbox.stub(utils, 'isAccountDeployed').resolves(true);
    const createAccountStub = sandbox.stub(createAccountUtils, 'createAccount');
```

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:144]:
```suggestion
    );
    expect(createAccountStub).to.have.been.calledOnceWith(sinon.match.any, true, true);
```

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:148]:
```suggestion
    sandbox.stub(utils, 'isAccountDeployed').resolves(false);
    const createAccountStub = sandbox.stub(createAccountUtils, 'createAccount').resolvesThis();
```

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:201]:
```suggestion
    );
    expect(createAccountStub).to.have.been.calledOnceWith(sinon.match.any, true, true);
```

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:205]:
```suggestion
    sandbox.stub(utils, 'isAccountDeployed').resolves(true);
```
make it deployed to bypass the mock of  the create account, as we have already verify from above test case

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:231]:
```suggestion
        { maxFee: '22702500105945', nonce: undefined },
```

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:241]:
```suggestion
sandbox.stub(utils, 'isAccountDeployed').resolves(true);
```
add back account deployed is true to bypass mock

**Review by stanleyyconsensys — CHANGES_REQUESTED** [review]
LGTM, some small changes suggestion on unit test

usually, when we test different Branches (if else then) of the code, we have to check if the statement inside the if has been triggered or not

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (khanti42) [reply packages/starknet-snap/test/src/executeTxn.test.ts:310]: @stanleyyconsensys thanks! I noticed I had forgotten on branch case (executeTxn multiple with account not deployed). I followed your suggestion in my last commit and also added two branches : 

`when account is not deployed` and `when account is deployed`


tests looks like this now : 

```
  Test function: executeTxn
    ✔ should throw error if executeTxn fail (63ms)
    ✔ should return false if user rejected to sign the transaction (59ms)
    when account is not deployed
      ✔ should executeTxn correctly and deploy an account (91ms)
      ✔ should executeTxn multiple and deploy an account (89ms)
    when account is not deployed
      ✔ should executeTxn and not deploy an account (63ms)
      ✔ should executeTxn multiple and not deploy an account (63ms)
```

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (stanleyyconsensys) [inline packages/starknet-snap/test/src/executeTxn.test.ts:88]: expect statement should not put in afterEach

we usually put some reset statement in afterEach

for determine statement, we put it inside the test, else, it is very hard to follow

and 

actually in MM style, they dont prefer sub test flow
Not:
when xxxx
    when yyyy
       it zzzz

Prefer:
when xxxx and yyyy
   it zzzz

i knew that some testing script is already using sub test flow, but would you mind to change it back

other than that, everything is good to go

**Review by stanleyyconsensys — CHANGES_REQUESTED** [review]

File: packages/starknet-snap/test/src/executeTxn.test.ts
Comment (khanti42) [reply packages/starknet-snap/test/src/executeTxn.test.ts:88]: ok let me roll back this change.

**Review by stanleyyconsensys — APPROVED** [review]

**Pull request merged**
